pkg/pool-weighted/contracts/smart/InvestmentPool.sol

```diff
@@ -85,22 +103,99 @@ contract InvestmentPool is BaseWeightedPool, ReentrancyGuard {
             owner
         )
     {
-        uint256 numTokens = tokens.length;
-        InputHelpers.ensureInputLengthMatch(numTokens, normalizedWeights.length, assetManagers.length);
+        uint256 totalTokens = tokens.length;
+        InputHelpers.ensureInputLengthMatch(totalTokens, normalizedWeights.length, assetManagers.length);
 
-        _tokenCountCache = numTokens;
+        _poolState = _poolState.insertUint7(totalTokens, _TOTAL_TOKENS_OFFSET);
 
-        // I'm time-traveling a bit here - storing the weights in a form where they can be changed
         uint256 currentTime = block.timestamp;
         _startGradualWeightChange(currentTime, currentTime, normalizedWeights, normalizedWeights, tokens);
+
+        // If false, the pool will start in the disabled state (prevents front-running the enable swaps transaction)
+        _setSwapEnabled(swapEnabledOnStart);
+    }
+
+    // External functions
+
+    /**
+     * @dev Tells whether swaps are enabled or not for the given pool.
+     */
+    function getSwapEnabled() public view returns (bool) {
+        return _poolState.decodeBool(_SWAP_ENABLED_OFFSET);
+    }
+
+    /**
+     * @dev Return start time, end time, and endWeights as an array.
+     * Current weights should be retrieved via `getNormalizedWeights()`.
+     */
+    function getGradualWeightUpdateParams()
+        external
+        view
+        returns (
+            uint256 startTime,
+            uint256 endTime,
+            uint256[] memory endWeights
+        )
+    {
+        // Load current pool state from storage
+        bytes32 poolState = _poolState;
+
+        startTime = poolState.decodeUint32(_START_TIME_OFFSET);
+        endTime = poolState.decodeUint32(_END_TIME_OFFSET);
+
+        (IERC20[] memory tokens, , ) = getVault().getPoolTokens(getPoolId());
+        uint256 totalTokens = tokens.length;
```

Comment:
I'm assuming there's a good reason, but wondering what it is. Why not use the cached `_getTotalTokens()` here?